### PR TITLE
ENH: retain masked EA dtypes in groupby with as_index=False

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/extension/json/test_json.py [hunk 1]
```diff
@@ -312,10 +312,6 @@ def test_groupby_extension_apply(self):
         we'll be able to dispatch unique.
         """
 
-    @pytest.mark.parametrize("as_index", [True, False])
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/extension/json/test_json.py:315]: it doesn't.  this deletion can be done independently of the rest of the PR

File: pandas/tests/extension/test_boolean.py [hunk 2]
```diff
@@ -269,14 +269,14 @@ def test_grouping_grouper(self, data_for_grouping):
...
 
-        index = pd.Index(index, name="B")
-        expected = pd.Series([3, 1], index=index, name="A")
         if as_index:
+            index = pd.Index(uniques, name="B")
+            expected = pd.Series([3, 1], index=index, name="A")
             self.assert_series_equal(result, expected)
         else:
-            expected = expected.reset_index()
+            expected = pd.DataFrame({"B": uniques, "A": [3, 1]})
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/extension/test_boolean.py:279]: yes

File: pandas/core/groupby/grouper.py [hunk 3]
```diff
@@ -601,6 +604,22 @@ def codes(self) -> np.ndarray:
         # expected "ndarray")
         return self._codes  # type: ignore[return-value]
 
+    @cache_readonly
+    def result_arraylike(self) -> ArrayLike:
+        """
+        Analogous to result_index, but holding an ArrayLike to ensure
+        we can can retain ExtensionDtypes.
+        """
+        ridx = self.result_index  # initialized _group_arraylike
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:613]: i agree, the statefulness is unpleasant.  #41375 starts to unwind it

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/groupby/grouper.py:613]: can you try to unwind first? this is adding a lot

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:613]: Sure, let’s get 41375 in and then I’ll rebased and try to trim this down

**Comment by jreback** [comment]
can you rebase and will look again

**Comment by jreback** [comment]
can you rebase again

**Comment by jbrockmendel** [comment]
just pushed.  this should be correct, but im also working on another branch that could serve as a state-reducing preliminary

**Comment by jorisvandenbossche** [comment]
Question: it's certainly nice to already preserve the dtype with `as_index=False`, but is it worth adding special code for it now if supporting EAs in the Index will also solve it? (assuming it does)

**Comment by jbrockmendel** [comment]
> Question: it's certainly nice to already preserve the dtype with as_index=False, but is it worth adding special code for it now if supporting EAs in the Index will also solve it? (assuming it does)

Good question.  Better to think of this as a disentangling/refactor of BaseGrouper/Grouping with a bonus upside of fixing the broken cases (in fact,  #41529 does most of this _except_ for fixing the as_index cases).  If/when we get the general EA Index , we'll be able to come back and trim this down some more.

**Comment by jbrockmendel** [comment]
gentle ping; id like to reclaim momentum on the groupby stuff

File: doc/source/whatsnew/v1.3.0.rst [hunk 4]
```diff
@@ -275,6 +275,7 @@ Other enhancements
 - Add keyword ``sort`` to :func:`pivot_table` to allow non-sorting of the result (:issue:`39143`)
 - Add keyword ``dropna`` to :meth:`DataFrame.value_counts` to allow counting rows that include ``NA`` values (:issue:`41325`)
 - :meth:`Series.replace` will now cast results to ``PeriodDtype`` where possible instead of ``object`` dtype (:issue:`41526`)
+- :class:`DataFrameGroupBy` operations with ``as_index=False`` now correctly retain ``ExtensionDtype`` dtypes for columns being grouped on (:issue:`41373`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:278]: can you move to 1.4

**Review by jreback — CHANGES_REQUESTED** [review]
can you rebase as well

File: doc/source/whatsnew/v1.3.0.rst (hunk 4 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.0.rst]: moved + green

File: pandas/core/groupby/ops.py [hunk 5]
```diff
@@ -1027,7 +1028,10 @@ def agg_series(
         else:
             result = self._aggregate_series_fast(obj, func)
 
-        npvalues = lib.maybe_convert_objects(result, try_float=False)
+        convert_datetime = obj.dtype.kind == "M"
```
Comment (jreback) [inline pandas/core/groupby/ops.py:1031]: what breaks if we remove this inference entirely?

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:1031]: huh, nothing now.  im pretty sure there was something back when i did this.  will revert

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: updated + green

**Comment by jreback** [comment]
might call this a bug fix

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
> might call this a bug fix

i think id rather call it an ENH, dont want users to think that other functions for which we havent yet done this are bugs

**Pull request merged**